slug.go

```diff
@@ -382,119 +398,98 @@ func (p *Packer) Unpack(r io.Reader, dst string) error {
 				// Create the symlink.
-				if err = os.Symlink(header.Linkname, path); err != nil {
+				if err = os.Symlink(header.Linkname, info.Path); err != nil {
 					return fmt.Errorf("failed creating symlink (%q -> %q): %w",
 						header.Name, header.Linkname, err)
 				}
 			} else {
 				return err
 			}
 
+			if err := info.RestoreInfo(); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		if info.IsDirectory() {
+			// Restore directory info after all files are extracted because
+			// the extraction process changes directory's timestamps.
+			directoriesExtracted = append(directoriesExtracted, info)
 			continue
 		}
 
-		// Only unpack regular files from this point on.
-		if header.Typeflag == tar.TypeDir || header.Typeflag == tar.TypeXGlobalHeader || header.Typeflag == tar.TypeXHeader {
+		// The remaining logic only applies to regular files
+		if !info.IsRegular() {
 			continue
-		} else if header.Typeflag != tar.TypeReg && header.Typeflag != tar.TypeRegA {
-			return fmt.Errorf("failed creating %q: unsupported type %c", path,
-				header.Typeflag)
 		}
 
 		// Open a handle to the destination.
-		fh, err := os.Create(path)
+		fh, err := os.Create(info.Path)
 		if err != nil {
 			// This mimics tar's behavior wrt the tar file containing duplicate files
 			// and it allowing later ones to clobber earlier ones even if the file
-			// has perms that don't allow overwriting.
+			// has perms that don't allow overwriting. The file permissions will be restored
+			// once the file contents are copied.
 			if os.IsPermission(err) {
-				os.Chmod(path, 0600)
-				fh, err = os.Create(path)
+				os.Chmod(info.Path, 0600)
+				fh, err = os.Create(info.Path)
 			}
 
 			if err != nil {
-				return fmt.Errorf("failed creating file %q: %w", path, err)
+				return fmt.Errorf("failed creating file %q: %w", info.Path, err)
 			}
 		}
 
-		// Copy the contents.
+		// Copy the contents of the file.
 		_, err = io.Copy(fh, untar)
 		fh.Close()
 		if err != nil {
-			return fmt.Errorf("failed to copy slug file %q: %w", path, err)
+			return fmt.Errorf("failed to copy slug file %q: %w", info.Path, err)
 		}
 
-		// Restore the file mode. We have to do this after writing the file,
-		// since it is possible we have a read-only mode.
-		mode := header.FileInfo().Mode()
-		if err := os.Chmod(path, mode); err != nil {
-			return fmt.Errorf("failed setting permissions on %q: %w", path, err)
+		if err := info.RestoreInfo(); err != nil {
+			return err
 		}
 	}
+
+	// Now that extraction is complete, restore mode and timestamps previously saved
+	// about directories. But first, sort the list of directories by lexical order, which
+	// should impose a top-down ordering so that, in case subdirectories appear first in
+	// the archive, they won't change the modified timestamps of their parent directories.
+	sort.Slice(directoriesExtracted, func(i, j int) bool {
+		return directoriesExtracted[i].Path < directoriesExtracted[j].Path
+	})
```

Comment:
I think this results in the wrong ordering, with `"foo"` before `"foo/bar"` which isn't what we want. Can we add a simple test for this?

Reply:
I don't think chtimes affects the parent directory mtime, because no matter how I order the tar headers for root and subdirectories, it doesn't seem to matter which order I restore the timestamps in. I created a test and a util to construct weird tar headers, but removed the sorting altogether.